lasagne/layers/noise.py

```diff
@@ -211,3 +219,69 @@ def get_output_for(self, input, deterministic=False, **kwargs):
             return input + self._srng.normal(input.shape,
                                              avg=0.0,
                                              std=self.sigma)
+
+
+class AlphaDropoutLayer(DropoutLayer):
+    """Dropout layer.
+
+    Sets values to alpha. Will also converge the remaining
+    neurons to mean 0 and a variance of 1 unit. See notes for disabling dropout
+    during testing.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        the layer feeding into this layer, or the expected input shape
+    p : float or scalar tensor
+        The probability of setting a value to zero
+    rescale : bool
+        If ``True`` (the default), scale the input by ``1 / (1 - p)`` when
+        dropout is enabled, to keep the expected output mean the same.
+    shared_axes : tuple of int
+        Axes to share the dropout mask over. By default, each value can be
+        dropped individually. ``shared_axes=(0,)`` uses the same mask across
+        the batch. ``shared_axes=(2, 3)`` uses the same mask across the
+        spatial dimensions of 2D feature maps.
+
+    Notes
+    -----
+    The alpha dropout layer is a regularizer that randomly sets input values to
+    zero and also applies a normalization function to bring the mean to 0
+    and the variance to 1; see [1]_for why this might improve
+    generalization. The behaviour of the layer depends on the ``deterministic``
+    keyword argument passed to :func:`lasagne.layers.get_output`. If ``True``,
+    the layer behaves deterministically, and passes on the input unchanged. If
+    ``False`` or not specified, dropout (and possibly scaling) is enabled.
+    Usually, you would use ``deterministic=False`` at train time and
+    ``deterministic=True`` at test time.
+
+    References
+    ----------
+    .. [1] Klambauer, G., Unterthiner, T., Mayr, A., Hochreiter, S. (2017):
+           Self-Normalizing Neural Networks. arXiv preprint: 1706.02515
+    """
+
+    def __init__(self, incoming, p=0.1, rescale=True, shared_axes=(),
+                 alpha=None, **kwargs):
+        """Class initialization."""
+        super(AlphaDropoutLayer, self).__init__(incoming,
+                                                p=p,
+                                                rescale=rescale,
+                                                shared_axes=shared_axes,
+                                                **kwargs)
+        if alpha is None:
+            self.alpha = -1.0507009873554804934193349852946 * \
+                         1.6732632423543772848170429916717
```

Comment:
You bring up good points. I will update the documentation accordingly. Regarding the option to pass a SELU instance and calculate the alpha based on that, I think that since AlphaDropout is meant to work with SELU, it only makes sense for it to be able to use SELU. I will make this change.